Add support for MAX30210

jemfgeronimo opened the pull request:
## PR Description

- This adds driver support, device tree bindings, and ABI documentation for added attributes for MAX30210 Temperature Sensor.
- Datasheet: [MAX30210](https://www.analog.com/media/en/technical-documentation/data-sheets/max30210.pdf)
- Tested on RPI4 with [MAX30210EVKIT](https://www.analog.com/en/resources/evaluation-hardware-and-software/evaluation-boards-kits/max30210evkit.html#eb-overview)
- Driver was initially done by Daniel Matyas: https://github.com/analogdevicesinc/linux/commit/8fc25e1cf70d12dabfb0622295fc2944337a369e
   - I've added features such as added events and attributes related to temperature rate-of-change
   - From what I've understood from his source code, I've also changed the trigger source of the triggered buffer from CVT/PDB pin to INT pin.

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

jemfgeronimo committed:
iio: Kconfig.adi: imply MAX30210

Add entry for the MAX30210 driver.

Signed-off-by: John Erasmus Mari Geronimo <johnerasmusmari.geronimo@analog.com>

jemfgeronimo committed:
iio: temperature: add driver for max30210

MAX30210 ±0.1°C Accurate Ultra-Small Low-Power Digital Temperature Sensor

Signed-off-by: John Erasmus Mari Geronimo <johnerasmusmari.geronimo@analog.com>

jemfgeronimo committed:
dt-bindings: iio: temperature: add adi,max30210.yaml

Add documentation for devicetree bindings for max30210

Signed-off-by: John Erasmus Mari Geronimo <johnerasmusmari.geronimo@analog.com>

jemfgeronimo commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+	.private = _private,						\
+}
+
+#define MAX30210_CHAN_EXT_INFO_RW(_name, _private)		\
```
This is unused. I will include this removal in the next set of changes.

nunojsa commented on Documentation/devicetree/bindings/iio/temperature/adi,max30210.yaml:
```diff
@@ -0,0 +1,105 @@
+  adi,comp-mode:
+    description: |
+      If present, comparator mode is used. If not present, interrupt mode is
+      used (default).
```
Do we need this property? Can't we assume this mode in the absence of an interrupt?

jemfgeronimo replied:
I adapted this from the existing code below: https://github.com/analogdevicesinc/linux/blob/5768e804cee9797437d8b39d6c11e94eafce20ce/Documentation/devicetree/bindings/iio/temperature/adi%2Cmax30210.yaml#L42
Comparator mode still uses interrupts. It just uses some register values differently, i.e., low temp alarm is used as high temp hysteresis.
This can also be controlled using debugfs_reg_access; should I just remove it?

jemfgeronimo replied:
I will use the following description:

      If present, the device is in comparator mode. In this mode, the low
      temperature threshold value acts as high temperature hysteresis. If not
      present, interrupt mode is used (default).

nunojsa commented on Documentation/devicetree/bindings/iio/temperature/adi,max30210.yaml:
```diff
@@ -0,0 +1,105 @@
+    description: |
+      If present, CVT/PDB pin is used for external temperature conversion. Value
+      is the active edge of CVT/PDB pin. If value is not present, 0 (falling
+      edge) is used (default).
```
This looks like a GPIO property? Moreover it seems that it can either be used as convert or powerdown pin. So we should likely reflect and support that

jemfgeronimo replied:
I only used this property to inform the driver how the device will use CVT/PDB pin so the driver will configure the proper registers for that but the control of the pin itself is done using an external source. Controlling the frequency of the pin if it is used as external convert pin is not done in this driver.
If this can be done differently, please let me know. Thanks!

nunojsa commented on Documentation/devicetree/bindings/iio/temperature/adi,max30210.yaml:
```diff
@@ -0,0 +1,105 @@
+    description: |
+      If present, value is the output drive type for the INT pin. Refer to
+      MAX30210 datasheet, bitfield INT_OCFG of the PIN CONFIGURATION (0x12) of
+      the Register Map. If value is not present, 0 is used (default).
```
Hmm at first glance this felt like it could be exported through gpiolib but I don't think we can control the level of the pin. Anyways, we should improve this binding by make it an enum and using strings to identify the possible controls.

nunojsa commented on Documentation/devicetree/bindings/iio/temperature/adi,max30210.yaml:
```diff
@@ -0,0 +1,105 @@
+      If present, temperature change detection is enabled. Value is the
+      Rate-of-Change FIR filter settings. Refer to MAX30210 datasheet, Table 5.
+      "Rate-of-Change Filter Settings." If value is not present, 0 is used
+      (default).
```
Take care about abusing referring the datasheet. If you then need to support more parts with this binding, things can quickly get dirty

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+static const unsigned long max30210_scan_masks[] = {
+	BIT(0),
+	0
+};
```
If we just have one channel isn't the above obvious?

jemfgeronimo replied:
should I just drop this and L1026:
```
1026 indio_dev->available_scan_masks = max30210_scan_masks;
```

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+	int ret;
+
+	if (!i2c_check_functionality(client->adapter, I2C_FUNC_SMBUS_BYTE_DATA |
+						      I2C_FUNC_SMBUS_I2C_BLOCK))
```
should align with open parenthesis. Also don't think you need the above check...

jemfgeronimo replied:
Is the reason here because I've used regmap and not i2c_smbus commands?

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+	if (IS_ERR(st->regmap))
+		return dev_err_probe(dev, PTR_ERR(st->regmap),
+					"Failed to allocate regmap.\n");
+
```
Before doing any ready, it's good practice to reset the device. I see there's a reset bit to do a SW register. I would also add support for the case where we have a powerdown pin.

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+					"Failed to read Part ID.\n");
+	if (val != MAX30210_PART_ID)
+		return dev_err_probe(dev, -ENODEV,
+					"Part ID mismatch.\n");
```
Typically now, we just warn about it and continue

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+				(int)ARRAY_SIZE(st->unique_id), st->unique_id);
+}
+
+static IIO_DEVICE_ATTR_RO(serialnumber, 0);
```
This is at most a debugfs attribute...

jemfgeronimo replied:
I have found this attribute documented below: https://github.com/analogdevicesinc/linux/blob/4b1dff42b7e7ccda802fb21684995653a47ab015/Documentation/ABI/testing/sysfs-bus-iio#L2205
For future reference, should I implement similar attributes as debugfs attributes?

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+				st->unique_id, ARRAY_SIZE(st->unique_id));
+	if (ret)
+		return dev_err_probe(dev, ret,
+					"Failed to read Serial Number.\n");
```
Just read this on demand and move it to a debugfs file

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+	ret = regmap_write(st->regmap, MAX30210_SYS_CONF_REG, 1);
+	if (ret)
+		return dev_err_probe(dev, ret,
+				      "Failed to reset the register values.\n");
```
The reset should be done before accessing the part...

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+	indio_dev->modes = INDIO_DIRECT_MODE;
+	indio_dev->available_scan_masks = max30210_scan_masks;
+
+	fwnode = dev_fwnode(dev);
```
drop the above

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+
+	fwnode = dev_fwnode(dev);
+
+	st->ext_cvt_en = fwnode_property_read_bool(fwnode, "adi,ext-cvt-active-edge");
```
Use device_property_read_bool(). Ditto for the other functions. Also, just directly use  device_property_read_u8() and check the ret value. If 0, the property is present and was read.

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+		}
+	}
+
+	fwnode_property_read_u8(fwnode, "adi,int-output-drive-type", &prop_val);
```
no return code being checked. The proper pattern is:

```
ret = device_property_read_u8(...);
if (!ret) {
    ...
}
```

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+		if (ret)
+			return ret;
+
+		indio_dev->trig = iio_trigger_get(st->trig);
```
no need to do the above...

jemfgeronimo replied:
when I drop this line, I cannot enable the buffer at runtime

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+
+		ret = devm_request_irq(dev, client->irq,
+					iio_trigger_generic_data_rdy_poll,
+					dir, indio_dev->name, st->trig);
```
why not threaded IRQ?

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+	MAX30210_CHAN_EXT_INFO_WO("roc_filter", CHAN_RATE_CHG_FILTER),
+	MAX30210_CHAN_EXT_INFO_RO("roc_length", CHAN_RATE_CHG_FILTER),
+	MAX30210_CHAN_EXT_INFO_RO("roc_en", CHAN_CHG_DET_EN),
+	{},
```
don't you have a roc filter binding? Why do you also need a runtime parameter? Also prepare to justify for this property

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+		}
+		ret = kstrtou32(buf, 10, &reg_val);
+		if (ret)
+			return ret;
```
can we accept any value? This is also just assuming raw values to be passed by userspace which have no meaning. I did not went into detail on the datasheet but what's the purpose of this rate of change? Is it oversampling, decimating? I suspect this can be implemented with standard ABI

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+		if (ret)
+			return ret;
+		return sysfs_emit(buf, "%d\n",
+				       sign_extend32(be16_to_cpu(uval), 8) * 5);
```
align on open paranthensis

jemfgeronimo replied:
should the 's' of 'sign_extend32' be directly under '(' or after that?

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+		if (!st->chg_det_en) {
+			dev_err(&indio_dev->dev, "Temp change detection is disabled.\n");
+			return -EINVAL;
+		}
```
If not possible, just do not expose this interface in the first place...

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+		return sysfs_emit(buf, "%d\n",
+				       max30210_temp_change_filter_length[val]);
+	case CHAN_CHG_DET_EN:
+		return sysfs_emit(buf, "%d\n", st->chg_det_en);
```
Drop the above... If you want to have it, put in debugFs. One should now if enabled or not this property

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+	__be16 uval;
+
+	switch (private) {
+	case CHAN_TEMP_SLOPE:
```
What is the purpose of this attr?

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+		if (ret)
+			goto err_dmode;
+
+		usleep_range(8000, 8300);
```
fsleep()

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+{
+	if (uval >= 0x9)
+		return 1;
+	else
```
redundant else

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+
+static int max30210_cnv_rate_div(unsigned int uval)
+{
+	if (uval >= 0x9)
```
why 9? (hint: comment could be useful)

nunojsa reviewed (commented):
Alright, I'm done for a first round of reviewing

jemfgeronimo commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1137 @@
+#define MAX30210_CHAN_EXT_INFO_WO(_name, _private) \
+	MAX30210_CHAN_EXT_INFO(_name, NULL, max30210_ext_info_write, _private)
+
+static const struct iio_chan_spec_ext_info max30210_ext_info[] = {
```
@nunojsa 
These attributes are related to the rate-of-change feature of the device. The idea is these attributes are accessible when rate-of-change mode of the device is enabled. I had second thoughts on adding these attributes because technically all of them can be accessed using debugfs_reg_access. Should I just remove them?

jemfgeronimo commented:
- Removed math from macros
- Fixed implementations of events that depends on change detect enable (roc-en) and alert mode (comp-int)
- Added PWM to implement the multifunction pin CVT/PDB
- Moved extra attributes to debugfs
- Made adi,int-output-drive-type into string
- Fixed indentions
- Added register reset in probe()
- Changed fwnode_property to device_property
- Added powerdown-gpios in dt-bindings
- Removed roc-filter from dt-bindings
- Renamed dt property adi,comp-mode to adi,comp-int

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+	} else {
+		prop_val = 0;
+	}
+	*dir = prop_val % 0x2 ? IRQF_TRIGGER_RISING : IRQF_TRIGGER_FALLING;
```
Maybe the above does not really need a dedicated DT property. You can use whatever one defined in DT for the IRQ pin. But I think it only makes sense if there's any dependency of the "output-drive-type" with the presence of an interrupt?

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+	if (prop)
+		st->events_type |= 0x2;
+	else
+		st->events_type &= ~(0x2);
```
Can we have both events enabled? Also, use __set_bit() instead... events_type will also be 0 initialized so I'm don't think you have to clear the bits...

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+{
+	struct iio_dev *indio_dev = arg;
+	struct max30210_state *st = iio_priv(indio_dev);
+	__be64 uval;
```
no DMA safe... the maintainer will complain about this

jemfgeronimo replied:
`__be64 uval __aligned(IIO_DMA_MINALIGN);`
Will that suffice? Thanks

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+{
+	struct iio_dev *indio_dev = arg;
+	struct max30210_state *st = iio_priv(indio_dev);
+	__be16 uval;
```
ditto about DMA safety...

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+	val = ilog2(val - 1);
+
+	return regmap_update_bits(st->regmap, MAX30210_TEMP_CONF_1_REG,
+				  MAX30210_RATE_CHG_FILTER_MASK, val);
```
Not sure if  I was the one suggesting this in the first place :sweat_smile:? But I'm not sure this is debugfs material? What is this roc_len about?

jemfgeronimo replied:
After giving it a second thought, I think it's more appropriately placed in DT. This can be considered similar to [fault-q](https://github.com/analogdevicesinc/linux/blob/0b3d96fbc075ccd36593cbed0825154e7ef8a17e/Documentation/devicetree/bindings/hwmon/adi%2Cmax31827.yaml#L49) as a number of samples, in this case, number of samples considered for rate-of-change.

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+static void max30210_fifo_read(struct iio_dev *indio_dev)
+{
+	struct max30210_state *st = iio_priv(indio_dev);
+	u8 data[3 * MAX30210_FIFO_SIZE];
```
No DMA safety

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+static int max30210_write_temp_raw(struct regmap *regmap, unsigned int reg,
+				    int temp)
+{
+	const __be16 uval = cpu_to_be16(temp);
```
no need for const and same comment about DMA safety

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+	if (ret)
+		return ret;
+
+	st->watermark = val;
```
This needs a lock... Also, should we be allowed to change the watermark while buffering? Would that work out?

jemfgeronimo replied:
If I remember correctly, the hwfifo_set_watermark runs every after preenable of the triggered buffer. I think st->watermark does not need a lock in this situation.

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+
+		return regmap_update_bits(st->regmap, MAX30210_TEMP_CONF_2_REG,
+					  MAX30210_TEMP_PERIOD_MASK,
+					  (unsigned int)data);
```
Should we be allowed to do this while buffering?

jemfgeronimo replied:
I will add `guard(mutex)(&st->lock);` at the start of `case IIO_CHAN_INFO_SAMP_FREQ:`

nunojsa commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+		if (ret)
+			goto err_dmode;
+
+		scoped_guard(mutex, &st->lock) {
```
Trying to understand if we need this guard... Are you protecting this code against what? The trigger handler? Asking this because you already have iio_device_claim_direct_mode() which guarantees the above code will not work while we're buffering (and hence running the trigger handler code)

jemfgeronimo replied:
I just recently discovered that iio_device_claim_direct_mode() and iio_device_release_direct_mode() are just mutex_lock() and mutex_unlock(). I'm wondering why I didn't check it before, but anyway, now my question is can I just use iio_device_claim_direct_mode() and iio_device_release_direct_mode() to all parts of my code that needs locking/unlocking? I just used guard(mutex) and scoped_guard(mutex,...) due to what I've seen in existing drivers.

nunojsa reviewed (commented):
Some stuff that still needs to be addressed but I would say that this is already in shape to be sent upstream.

jemfgeronimo commented on MAINTAINERS:
```diff
@@ -12587,6 +12587,16 @@ F:	Documentation/userspace-api/media/drivers/max2175.rst
+W:	https://ez.analog.com/linux-software-drivers
+F:	Documentation/ABI/testing/sysfs-bus-iio-temperature-max30210
+F:	Documentation/devicetree/bindings/iio/temperature/adi,max30210.yaml
+F:	drivers/iio/temperature/max30210.c
```
Got it! Thanks!

jemfgeronimo commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to read Part ID.\n");
+	if (val != MAX30210_PART_ID)
+		dev_err_probe(dev, -ENODEV, "Part ID mismatch.\n");
```
Noted, thanks!

machschmitt replied:
Don't know how maintainers are considering warnings from IIO devices but, not long ago, Greg K-H said to avoid kernel warnings https://www.youtube.com/watch?v=Rg_VPMT0XXw.
In sum, everything that can trigger a dev_warn() would be considered a kernel CVE.
If *panic_on_warn* is enabled, reboot would happen.

jemfgeronimo commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1218 @@
+	if (!device_property_read_string(dev, "adi,int-output-drive-type",
+					 &str_val)) {
+		prop_val = sysfs_match_string(max30210_int_drive_types,
+					      str_val);
```
Thanks for this info! :)

jemfgeronimo commented:
V3
- remove description of `reg` property
- remove `adi,int-output-drive-type` property
- remove `roc_length` from debugfs & add `adi,roc-len` property
- improve description of `adi,comp-int` property
- remove Daniel Matyas name in `MAINTAINERS` file
- add `__aligned(IIO_DMA_MINALIGN)`
- add `iio_device_claim_direct_mode()` in `IIO_CHAN_INFO_SAMP_FREQ` of `max30210_write_raw()`
- improvement of PWM implementation
- add use of `__set_bit()`
- used `switch-case` instead of `if-else` in some cases
- fix typo of `max30210_debugfs_init()`

jemfgeronimo commented on drivers/iio/temperature/max30210.c:
```diff
@@ -0,0 +1,1208 @@
+static int max30210_read_temp_raw(struct regmap *regmap, unsigned int reg,
+				   int *temp)
+{
+	u8 uval[2] __aligned(IIO_DMA_MINALIGN);
```
I can do that. Or can I just remove `__aligned(IIO_DMA_MINALIGN)`? I honestly don't know the use cases of this alignment. I will appreciate an explanation. Thanks.

machschmitt replied:
Not an expert on DMA, but I don't think removing the alignment pragma would be a good thing. I think it is likely that upstream IIO maintainer will ask for explicit buffer alignment to enforce DMA safety. IIRC, the SPI controller in some platforms can talk to the DMA controller and ask SPI transfers to be DMA mapped (rpi can do this IIRC). That would happen seamlessly to the device driver. In those cases, the buffer data can get corrupted if it's not properly aligned on cache lines or miss-handled by the DMA controller. There are other situations in which buffers can get corrupted when data moves along through DMA. Some upstream content about DMA buffer alignment:
https://lore.kernel.org/linux-iio/20240413173409.63d33a0a@jic23-huawei/
https://lore.kernel.org/all/20220508175712.647246-1-jic23@kernel.org/
https://lore.kernel.org/all/20220508175712.647246-2-jic23@kernel.org/
Anyway, I'm not expert on DMA so other reviewers may provide better guided suggestions here.

nunojsa commented:
@jemfgeronimo, What is the status of this PULL?

nunojsa commented:
Hi, I'm just checking in on the status of this pull request. Is there anything I can do to help move it forward?
